minetest/minetest
# Clip posteffect color to respective nodes

* Fixes #12091.
* Fixes #12429.

### Screenshots

<details>

![black_ceiling](https://github.com/user-attachments/assets/3dd03a03-4b8e-4fa6-9a52-f3845436001b)
![waving_water](https://github.com/user-attachments/assets/4886d437-0f55-4c30-b95b-a935fa269a6b)
![water_and_black](https://github.com/user-attachments/assets/e82b0ad0-ca31-4669-b59b-3233b3c5f950)
![flowing_water](https://github.com/user-attachments/assets/c980092d-8e75-49ef-a6c4-30529c366a33)
![flowing_water_below](https://github.com/user-attachments/assets/bcae3713-86f1-407e-9a59-157d3759879a)
![flowing_water_down](https://github.com/user-attachments/assets/247a91f9-29f6-401f-a010-7f60eee66e42)
If there's no backface:
![no_backface](https://github.com/user-attachments/assets/90e54522-0887-493c-8c24-34e8239cbf73)
Posteffect fullscreen plus shaped:
![overlap](https://github.com/user-attachments/assets/ca88c889-d1cc-4abd-99fc-ad6513455abc)
A thin line between:
![pxlline1](https://github.com/user-attachments/assets/54f1af9c-237c-40b0-bbcb-b5cbd37e7b28)
![pxlline2](https://github.com/user-attachments/assets/5841341d-1a48-4633-9ffe-ab5147cd1aad)
Torch triggers different triangulation of water (not new in this PR. but PR needs to take this into account):
![torch_next_to_water](https://github.com/user-attachments/assets/40ff3f02-c2b7-45c4-a910-92956f7ce856)


</details>

### How it works

#### In master
For comparison, in master, the camera position is used to determine node whose post effect color should be used. Then a rectangle is drawn over the whole screen. It is drawn after the 3D scene and after the wield hand.

#### In PR
For PR, imagine how the near plane lies in the world. Because it is small, it will only intersect with nodes in a 2x2x2 block around the camera pos.
If the near plane intersects with a node, the respective are on the screen is inside the node, and needs posteffect. We want to draw a shape with the posteffect color for this area.

#### Wieldhand
If we drew the wieldhand before the posteffect shapes, the wieldhand would often be partly covered by such a shape, which is ugly. This is why the wieldhand is rendered after the posteffect step.
To color the wield hand appropriately, there's a new uniform for the object shader (used for wieldhand).
If the whole screen is filled with one posteffect color, it looks exactly like in master.
If shaders are off, the node light color is abused. It looks different. But shaders off support will be dropped soonish anyways.

#### How the shapes are determined
Imagine a node with drawtype normal for now, it is just a cube. More generally it is a convex polytope. A convex polytope is just an intersection of a bunch of half spaces, given via oriented planes, which are represented by 4D vectors.
These planes are transformed to clip space, and the intersection with the near plane (an oriented 2D line) is gotten (see `clip_polygon_by_objsp_plane()`).
To get the shape, we start with a rectangle filling the whole screen, and then cut off parts of it with one of the oriented 2D lines from above (see `ConvexPolygon::clip()`). The result (and all steps in between) are a convex polygon.
In the general case, we have more than one convex polytope for a node, and end up with multiple convex polygons.

#### A note about liquids
Liquids (e.g. water) can wave. The vertices at the top are moved up and down. It's still just two convex polygons. We cut the node by it's top face quad diagonal, which depends on lighting. There's also flowing liquid. Most of the changed lines outside of `clientmap.cpp` are devoted to either exposing the necessary information for this, or to coloring the wieldhand.

#### Other drawtypes
For the sake of backwards compatibility, other drawtypes than flowing liquid, liquid source, and normal still work similar as in master: They fill the whole screen and are only drawn if and only if the camera is inside the node.
This means that multiple post effect polygons can overlap. Check out the screenshots to see how this looks like.


### Known issues

* There's sometimes a 1 pixel or so thick line between the posteffect shape and the node face. (See screenshots.)
* ~~The wield shading doesn't work correctly without shaders.~~ We no longer support the fixed function pipeline.
* In cases where #14159 appears, the posteffect shape doesn't behave like the buggy water.
* Far out in the map, waving may also be broken, near `/teleport 30140 0 10` for example. (Idk why.)
* Extreme water wave parameters don't work.
* If you're in a node that has backface culling, it can look a bit weird, because you can see the border of the node without seeing the face. But imo it looks correct, and is fine. (See screenshots.)

### Possible future PRs

* Add API for nodes to define convex polytopes to use for the posteffect.

## To do

This PR is ready for review.

Things I maybe should improve:
* Respect undersampling setting. (Render the 2d things with low resolution.)
* I think I can make the uniform in the wield (object) fragment shader now optional.
* Try to set floating point flags to make it match what my gpu does for waving liquids.
  => Used `_mm_setcsr` to enable DNZ and FTZ flags in `ClientMap::getLiquidTopFaceHeights`, didn't appear to reproduce what the gpu does.

## How to test

* Fly inside a node, then disable noclip. Or in case of water, just fly into it.
* Settings to try: `enable_shaders`, `enable_waving_water`, `water_wave_*`, `3d_mode`, noclip, `smooth_lighting`, camera mode (1st pers, 2nd pers, 3rd pers), `video_driver`
* Try at different positions in the world.
* Try with different nodes / node combinations, including solid next to liquid (e.g. so it doesn't wave below), node with fullscreen posteffect color, and waving liquid.
* Place a light source to trigger different triangulation of liquid top. (See screenshots.)
* Go underwater and take a screenshot of the wieldhand. Compare colors to master.
* Here's a mod for an allfaces node with opaque green posteffect color: https://codeberg.org/Desour/test_tmp/src/branch/posteffect_green (notice the branch)
